lib/SyTest/Federation/Datastore.pm

```diff
@@ -190,27 +190,30 @@ sub create_event
       for qw( type auth_events content depth prev_events room_id sender );
 
    my $room_version = delete $fields{room_version} // 1;
+   my $event_id_suffix = delete $fields{event_id_suffix};
 
    my $event = {
       %fields,
       origin           => $self->server_name,
       origin_server_ts => JSON::number( int( time() * 1000 )),
    };
 
-   my $event_id = delete $fields{event_id};
-   my $suffix = delete $fields{event_id_suffix};
-   if( not defined $event_id ) {
-      if( $room_version eq '1' || $room_version eq '2' ) {
-         # assign an event id
-         $event_id = $self->next_event_id( $suffix );
+   my $event_id = $fields{event_id};
+   if( $room_version eq '1' || $room_version eq '2' ) {
+      if( not defined $event_id ) {
+         # room v1/v2: assign an event id
+         $event_id = $self->next_event_id( $event_id_suffix );
          $event->{event_id} = $event_id;
-      } else {
-         # calculate the event id from the hash
-         $event_id = id_for_event( $event, $room_version );
       }
+      $self->sign_event( $event );
+   } else {
+      die "event with explicit event_id in room v$room_version"
+         if defined $event_id;
+
+      $self->sign_event( $event );
+      $event_id = id_for_event( $event, $room_version );
```

Comment:
Isn't this now calculating the hash after signing?

Reply:
that's what I (wish I) said

Reply:
Why does it matter if we hash before or after we sign? The signatures should be stripped before we calculate the hash?

Reply:
yes, but the `hashes` aren't.

Reply:
Oh, and `sign_event` also adds the hash. Right, got you. So this basically says we need to calculate the event ID hash *after* we've added the content hashes.